src/MSBuildBinLogQuery/Token/Token.cs

```diff
@@ -0,0 +1,6 @@
+namespace Microsoft.Build.Logging.Query.Token
+{
+    public abstract class Token
+    {
```

Comment:
Consider internal constructor here to express that the set of tokens is not open ended. All derived classes are in this assembly and sealed. Really, this is a discriminated union expressed as an object-oriented hierarchy.

Depending on how you expect this language to evolve, you might also consider expressing the discriminated union as a single struct Type with a Kind enum and string Text value. It's less pretty, but would further cut down on heap allocations, and switching over the kind is going to be a lot faster than doing type checks.

Reply:
Yeah, I actually considered a `Kind` enum and a string `Text` value. But C# doesn't have a `Void` or `None` type which is awkward...

Reply:
> But C# doesn't have a Void or None type which is awkward...

How is that related to the choice of a struct with Kind and Text vs. the class hierarchy. For the hierarchy, you are using `null` as your "none". You could do that with the struct too, if you used `Token?`. Or you could have a None kind.

Reply:
I think what I want here is a class `Token<TValue>` with the type parameter `TValue`. So for the string token it would be `Token<string>` and for the integer token (to be added) it would be `Token<int>`. For others, ideally I want `Token<None>` or something similar (like `Void` in Java), but C# doesn't have this.